### PR TITLE
Ensure we dedup normalizers

### LINKED ISSUE #1676: Dedup normalizers
If one has a range and a given field has multiple normalizers, we are not dedupping the values if the normalizers all produce the same output. We could improve query performance by ensuring we dedupe bounded ranges with multiple normalizers that end up producing the same wokr.

### PR DESCRIPTION
**Pull request by mineralntl**

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java [hunk 1]
```diff
@@ -188,7 +185,21 @@ private Object expandRangeForNormalizers(LiteralRange<?> range, JexlNode node) {
                 continue;
             }
             
-            aliasedBounds.add(new BoundedRange(JexlNodes.children(new ASTAndNode(ParserTreeConstants.JJTANDNODE), left, right)));
+            // check to ensure we're not adding duplicate ranges with identical values
+            BoundedRange newAliasedBound = new BoundedRange(JexlNodes.children(new ASTAndNode(ParserTreeConstants.JJTANDNODE), left, right));
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: It seems to me that we could implement `equals()` and `hashcode()` for the BoundedRange object using some of the string building logic here. 

Then line 159 could change from `List<BoundedRange>` to `Set<BoundedRange>` and this extra logic goes away.

Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: Updated as per suggested

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/BoundedRange.java [hunk 2]
```diff
@@ -54,11 +56,27 @@ public BoundedRange(JexlNode node) {
...
+        }
+        if (!(obj instanceof BoundedRange)) {
+            return false;
+        }
+        return JexlStringBuildingVisitor.buildQuery(this).equals(JexlStringBuildingVisitor.buildQuery((BoundedRange) obj));
+    }
+    
+    @Override
+    public int hashCode() {
+        return new HashCodeBuilder().append(JexlStringBuildingVisitor.buildQuery(this)).toHashCode();
```
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/jexl/nodes/BoundedRange.java]: I missed this in the initial review, these calls should be `buildQueryWithoutParse`. Otherwise this merge request is good to go, pending integration testing

**Review by apmoriarty — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/BoundedRange.java (hunk 2 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/nodes/BoundedRange.java]: Updated as requested

**Review by apmoriarty — DISMISSED** [review]

**Review by wgtong — DISMISSED** [review]
You'll probably have to update some tests on the hs as well

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java [hunk 3]
```diff
@@ -198,16 +195,16 @@ private Object expandRangeForNormalizers(LiteralRange<?> range, JexlNode node) {
             
             // Avoid extra parens around the expansion
             if (1 == aliasedBounds.size()) {
-                return aliasedBounds.get(0);
+                // we assume the Optional container is not empty as we have already created a new BoundedRange from the node
+                return Arrays.stream(aliasedBounds.toArray()).findFirst().get();
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: Much more efficient: return aliasedBounds.iterator().next();

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java [hunk 4]
```diff
@@ -198,16 +195,16 @@ private Object expandRangeForNormalizers(LiteralRange<?> range, JexlNode node) {
...
             } else {
-                List<ASTReferenceExpression> var = JexlASTHelper.wrapInParens(aliasedBounds);
+                List<ASTReferenceExpression> var = JexlASTHelper.wrapInParens(aliasedBounds.stream().collect(Collectors.toList()));
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: Probably more efficient: new ArrayList(aliasedBounds);

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java [hunk 5]
```diff
@@ -198,16 +195,16 @@ private Object expandRangeForNormalizers(LiteralRange<?> range, JexlNode node) {
             
             // Avoid extra parens around the expansion
             if (1 == aliasedBounds.size()) {
-                return aliasedBounds.get(0);
+                // we assume the Optional container is not empty as we have already created a new BoundedRange from the node
+                return aliasedBounds.iterator().next();
             } else {
-                List<ASTReferenceExpression> var = JexlASTHelper.wrapInParens(aliasedBounds);
+                List<ASTReferenceExpression> var = new ArrayList(aliasedBounds);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: You still need the JexlASTHelper.wrapInParens() part...

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java (hunk 3 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: Updated as requested

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java (hunk 5 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: I'll look into adding a few other unit tests because fun fact: this change passed all tests before pushing.

Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: I think the parens are required for a few edge cases, so I could believe that.

File: warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTermsTest.java [hunk 6]
```diff
@@ -237,6 +239,38 @@ public void testBoundedMultiNormalizedBounds() throws ParseException {
...
+        
+        config.setQueryFieldsDatatypes(dataTypes);
+        
+        String original = "((_Bounded_ = true) && (FOO > 1 && FOO < 10))";
+        String expected = "((_Bounded_ = true) && (FOO > '1' && FOO < '10'))";
+        expandTerms(original, expected);
+    }
+    
+    @Test
+    public void testBoundedMultiNormalizedBounds3() throws ParseException {
```
Comment (hlgp) [inline warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTermsTest.java:259]: nice tests

Comment (mineralntl) [reply warehouse/query-core/src/test/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTermsTest.java:259]: 👍

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java (hunk 4 above)
Comment (mineralntl) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ExpandMultiNormalizedTerms.java]: Updated in last commit

**Review by hlgp — APPROVED** [review]

**Review by apmoriarty — APPROVED** [review]

**Review by ivakegg — APPROVED** [review]
